### PR TITLE
[SPARK-33162][INFRA] Use pre-built image at GitHub Action PySpark jobs

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to use `pre-built image` at Github Action PySpark jobs. To isolate the changes, `pyspark` jobs are split from the main job. The docker image is built by the following.

| Item                   | URL                |
| --------------- | ------------- |
| Dockerfile         | https://github.com/dongjoon-hyun/ApacheSparkGitHubActionImage/blob/main/Dockerfile |
| Builder               | https://github.com/dongjoon-hyun/ApacheSparkGitHubActionImage/blob/main/.github/workflows/build.yml |
| Image Location | https://hub.docker.com/r/dongjoon/apache-spark-github-action-image |

Please note that.
1. The community still will use `build_and_test.yml` to add new features like as we did until now. The `Dockerfile` will be updated regularly.
2. When Apache Spark gets an official docker repository location, we will use it.
3. Also, it's the best if we keep this docker file and builder script at a new Apache Spark dev branch instead of outside GitHub repository.

### Why are the changes needed?

Currently, two `pyspark` test jobs take over one and half hour always. In total, 3 hours 14 minutes.
- https://github.com/apache/spark/runs/1240470628 (1 hour 35 mins)
- https://github.com/apache/spark/runs/1240470634 (1 hour 39 mins)

This PR will remove the package installation steps which takes 16 minutes and causes flakiness. Note that `Python 3.6 package installation` is not included in the pre-built image and it only takes `20s`.

**BEFORE**
Screen Shot 2020-10-15 at 10 32 17 AM

**AFTER**
Screen Shot 2020-10-15 at 10 58 17 AM

In short, `pyspark` GitHub jobs take shorter time. In total, 2 hours 23 minutes (<- 3 hours 14 minutes, previously).
- https://github.com/apache/spark/pull/30059/checks?check_run_id=1260512568 (1 hour 18 mins)
- https://github.com/apache/spark/pull/30059/checks?check_run_id=1260512582 (1 hour 5 mins)

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass the GitHub Action on this PR without `package installation steps`.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Could you review this please, @srowen , @HyukjinKwon , @maropu , @viirya ?

**Review by srowen — COMMENTED** [review]
Can the Dockerfile etc live in the Spark project?
Not a big deal.

**Comment by dongjoon-hyun** [comment]
Ya. In addition to that, I want to move forward to the following direction in the near future.
1. Apply the official Apache dockerhub location for Apache Spark
2. Make a new branch inside Apache Spark repo and move `Dockerfile` and `Github Action build file` by moving `https://github.com/dongjoon-hyun/ApacheSparkGitHubActionImage`'s main branch to that Apache Spark branch.

**Comment by holdenk** [comment]
One quick question: the docker image doesn't seem to install python3.7?
Can you add a USER to the dockerfile to specify a non-root user?

**Comment by dongjoon-hyun** [comment]
@holdenk . 
- This is a migration for AS-IS `Github Action` jobs. We don't test `Python 3.7` at all.
- I tried already `USER` but it's not-trivial work in Github Action. We still is using `GitHub Actions` like `actions/checkout@v2` which assumes `root`. If you are search that, there are multiple threads there.

**Comment by dongjoon-hyun** [comment]
BTW, this PR is irrelevant to `root` user issue as you see here.

File: .github/workflows/build_and_test.yml [hunk 1]
```diff
@@ -128,41 +124,17 @@ jobs:
...
-        python3.6 -m pip install numpy pyarrow pandas scipy xmlrunner
-        python3.6 -m pip list
-        # PyPy does not have xmlrunner
-        pypy3 -m pip install numpy pandas scipy
-        pypy3 -m pip list
     - name: Install Python packages (Python 3.8)
-      if: contains(matrix.modules, 'pyspark') || (contains(matrix.modules, 'sql') && !contains(matrix.modules, 'sql-'))
+      if: (contains(matrix.modules, 'sql') && !contains(matrix.modules, 'sql-'))
       run: |
         python3.8 -m pip install numpy pyarrow pandas scipy xmlrunner
```
Comment (viirya) [inline .github/workflows/build_and_test.yml:139]: Are these packages still necessary for sql related tests?

**Review by viirya — COMMENTED** [review]
As the pre-built image cannot be easily changed like modifying workflow yml file, the project contributors might be a bit harder to change Github Action PySpark job.

File: .github/workflows/build_and_test.yml (hunk 1 above)
Comment (dongjoon-hyun) [reply .github/workflows/build_and_test.yml:139]: Hi, @viirya 
`Python 3.8` is not the target of this PR because it doesn't consume any time (it's zero seconds).

Screen Shot 2020-10-15 at 3 22 48 PM

In addition to that, I want to keep it here without touching it. When we switch `Scala/Java` tests into `pre-built image`, it will use the shared python libraries on the image in any way.

**Comment by dongjoon-hyun** [comment]
Ur, why do you think like that? Our plan is [here](url).

1. If you are saying about `repo` or `branch`. The project contributors can change PySpark job inside Apache Spark repo in the future. It's the same cost like changing `branch-2.4` or `branch-3.0`. And, it's easier than changing `spark-website`. 
> As the pre-built image cannot be easily changed like modifying workflow yml file, the project contributors might be a bit harder to change Github Action PySpark job.

2. If you are saying about `Dockerfile`, we already have 7 `Dockerfile` which is open for contributor. It consists of easy linux commands.
```
$ find . -name Dockerfile
./resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile
./resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/python/Dockerfile
./resource-managers/kubernetes/docker/src/main/dockerfiles/spark/bindings/R/Dockerfile
./dev/create-release/spark-rm/Dockerfile
./external/docker/spark-test/master/Dockerfile
./external/docker/spark-test/worker/Dockerfile
./external/docker/spark-test/base/Dockerfile
```

BTW, I want to focus on the improvement which we can get. As you know, I tried various approach in the following PR already, but nothing is better than this. This PR reduces from `3 hours 14 minutes` to `2 hours 23 minutes`.
- https://github.com/apache/spark/pull/30012

File: .github/workflows/build_and_test.yml (hunk 1 above)
Comment (dongjoon-hyun) [reply .github/workflows/build_and_test.yml:139]: Correction: It can take `20s`.

**Comment by viirya** [comment]
Because currently you can easily change workflow yml files and see what test result can be immediately. By switching to pre-built image, we might need change docker file first, publish updated image, then verify updated image work as expected with GitHub Actions workflow.

So for contributors, when they want to change something in the workflow, they might need to change docker file, update image at their dockerhub location, then use their custom image in workflow file.

This should be a minor concern because I believe we won't frequently change workflow stuffs. But just want to raise it explicitly.

**Comment by dongjoon-hyun** [comment]
Got it. I agree with you at those points.

**Comment by dongjoon-hyun** [comment]
BTW, you don't need to put everything into `Dockerfile`.
Let say, it's the same with the current one we are using `ubuntu:20.04`. Like you described, the contributor don't need to know about `pre-built image`. Instead, they can add new commands on GitHub Workflow file only. Then, it will be installed on the fly. And, we can move it at every release preparation. How do you think about this?

**Comment by dongjoon-hyun** [comment]
This PR doesn't block users to use `.github/workflows/build_and_test.yml` directly. The main benefit is just boosting the performance by removing repeated and flaky installation tasks.

**Comment by viirya** [comment]
Yes, I thought about it too. Except for very few things it is necessarily to change to Docker file, other changes can be in workflow file first. So it is just minor concern.

File: .github/workflows/build_and_test.yml [hunk 2]
```diff
@@ -201,6 +173,92 @@ jobs:
         name: unit-tests-log-${{ matrix.modules }}-${{ matrix.comment }}-${{ matrix.java }}-${{ matrix.hadoop }}-${{ matrix.hive }}
         path: "**/target/unit-tests.log"
 
+  pyspark:
+    name: "Build modules: ${{ matrix.modules }}"
+    runs-on: ubuntu-20.04
+    container:
+      image: dongjoon/apache-spark-github-action-image:20201015
```
Comment (viirya) [inline .github/workflows/build_and_test.yml:180]: As we have `container`, do we still need `runs-on`?

Comment (dongjoon-hyun) [reply .github/workflows/build_and_test.yml:180]: Yes. It's required. Many people complains about that. :)
- https://github.community/t/confused-with-runs-on-and-container-options/16258

Comment (viirya) [reply .github/workflows/build_and_test.yml:180]: interesting... :)

Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:180]: @dongjoon-hyun, quick question. what about we just set this image for the main job above (meaning use it for all tests and build), and remove this new job here? I think that'll be easier to move to use Docker image for all test cases.

File: .github/workflows/build_and_test.yml [hunk 3]
```diff
@@ -201,6 +173,92 @@ jobs:
...
+        key: pyspark-ivy-${{ hashFiles('**/pom.xml', '**/plugins.sbt') }}
+        restore-keys: |
+          pyspark-ivy-
+    - name: Install Python 3.6
+      uses: actions/setup-python@v2
+      if: contains(matrix.modules, 'pyspark')
+      with:
+        python-version: 3.6
+        architecture: x64
+    - name: Install Python packages (Python 3.6)
```
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:238]: Another quick question. So dose the docker image has Python 3.8 and PyPy3 with the packages installed? I wonder if we can install them about Python 3.6 as well.

File: .github/workflows/build_and_test.yml (hunk 2 above)
Comment (dongjoon-hyun) [reply .github/workflows/build_and_test.yml:180]: @HyukjinKwon . I tried already. Please see https://github.com/apache/spark/pull/30059#issuecomment-709555763 .

File: .github/workflows/build_and_test.yml (hunk 3 above)
Comment (HyukjinKwon) [reply .github/workflows/build_and_test.yml:238]: I mean, I got that it takes less time but wondering if we can just pre-install it to make it look consistent.
BTW, it's interesting that installing `numpy pyarrow pandas scipy xmlrunner` only takes 20s seconds ..

File: .github/workflows/build_and_test.yml (hunk 2 above)
Comment (dongjoon-hyun) [reply .github/workflows/build_and_test.yml:180]: That's the reason why I refocus on this only which is most time-saving one.
We can revisit the others later.

File: .github/workflows/build_and_test.yml (hunk 3 above)
Comment (dongjoon-hyun) [reply .github/workflows/build_and_test.yml:238]: There is a technical issue to install `Python 3.6` on `Ubuntu 20.04`. The following is the docker image.
- https://github.com/dongjoon-hyun/ApacheSparkGitHubActionImage/blob/main/Dockerfile

And, `Python 3.6` installation and package installation only takes 30 seconds. It's not big.

**Comment by dongjoon-hyun** [comment]
BTW, @HyukjinKwon . If you are interested,
- SparkR Prebuilt Testing is on the way here (https://github.com/dongjoon-hyun/spark/pull/37)
- Initial Container try was done here (https://github.com/dongjoon-hyun/spark/pull/35).

File: .github/workflows/build_and_test.yml (hunk 3 above)
Comment (HyukjinKwon) [reply .github/workflows/build_and_test.yml:238]: Gotya.

File: .github/workflows/build_and_test.yml (hunk 2 above)
Comment (HyukjinKwon) [reply .github/workflows/build_and_test.yml:180]: 👌

File: .github/workflows/build_and_test.yml (hunk 3 above)
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:238]:
```suggestion
    # This step takes much less time (~30s) than other Python versions so it is not included
    # in the Docker image being used. There is also a technical issue to install Python 3.6 on
    # Ubuntu 20.04. See also SPARK-33162.
    - name: Install Python packages (Python 3.6)
```

**Comment by dongjoon-hyun** [comment]
Thanks, @HyukjinKwon . I applied your suggestion.

File: .github/workflows/build_and_test.yml [hunk 4]
```diff
@@ -201,6 +173,95 @@ jobs:
...
+      uses: actions/cache@v2
+      with:
+        path: ~/.ivy2/cache
+        key: pyspark-ivy-${{ hashFiles('**/pom.xml', '**/plugins.sbt') }}
+        restore-keys: |
+          pyspark-ivy-
+    - name: Install Python 3.6
+      uses: actions/setup-python@v2
+      if: contains(matrix.modules, 'pyspark')
+      with:
```
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:235]: Nit but .. I think we can add this condition when we actually add some combinations of other modules. We don't know if we'll fix this job or fold this one to the main job above, also considering the name of the current job is `pyspark`.

```suggestion
      with:
```

File: .github/workflows/build_and_test.yml [hunk 5]
```diff
@@ -201,6 +173,95 @@ jobs:
...
+        python-version: 3.6
+        architecture: x64
+    # This step takes much less time (~30s) than other Python versions so it is not included
+    # in the Docker image being used. There is also a technical issue to install Python 3.6 on
+    # Ubuntu 20.04. See also SPARK-33162.
+    - name: Install Python packages (Python 3.6)
+      if: contains(matrix.modules, 'pyspark')
+      run: |
```
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:243]:
```suggestion
      run: |
```

**Review by HyukjinKwon — APPROVED** [review]
LGTM

File: .github/workflows/build_and_test.yml (hunk 4 above)
Comment (dongjoon-hyun) [reply .github/workflows/build_and_test.yml:235]: Sure. Thanks.

**Comment by dongjoon-hyun** [comment]
Thank you, @HyukjinKwon and all.
Merged to master!

**Comment by HyukjinKwon** [comment]
Oh btw, make sure the author is you @dongjoon-hyun.

**Comment by HyukjinKwon** [comment]
the number of my commits will suggest me as the main author in this PR in the merge script.

**Comment by HyukjinKwon** [comment]
Nice!

**Comment by HyukjinKwon** [comment]
That's completely no problem :-)

**Comment by dongjoon-hyun** [comment]
Ah.. GitHub Suggestion is making you as a committer instead of author.. Weird..

**Comment by HyukjinKwon** [comment]
Ah .. looks they changed their behaviour on that. No problem!

**Comment by dongjoon-hyun** [comment]
Sorry, I'll adjust by myself next time. :) Thank you for your suggestion!
